src/osd/PeeringState.cc

```diff
@@ -403,13 +403,49 @@ void PeeringState::write_if_dirty(ObjectStore::Transaction& t)
   }
 }
 
+#ifdef WITH_SEASTAR
+seastar::future<> PeeringState::advance_map(
+#else
 void PeeringState::advance_map(
+#endif
   OSDMapRef osdmap, OSDMapRef lastmap,
   vector<int>& newup, int up_primary,
   vector<int>& newacting, int acting_primary,
   PeeringCtx &rctx)
 {
   ceph_assert(lastmap == osdmap_ref);
+#ifdef WITH_SEASTAR
+  return seastar::do_with(std::move(newup), std::move(newacting),
+    [this, up_primary, acting_primary, osdmap, lastmap, &rctx]
+    (auto& newup, auto& newacting) {
+    return [this, &newup, &newacting, up_primary, acting_primary,
+	    osdmap, lastmap, &rctx] {
+      if (should_restart_peering(up_primary, acting_primary,
+				 newup, newacting, lastmap,
+				 osdmap)) {
+	return pl->pre_state_reset(pg_log);
```

Comment:
If we must do this, then so be it, but I'd rather not have a significant block of code in PeeringState duplicated for crimson.  Did you consider reactoring the PeeringState implementation and interface to avoid this?